File: ruby/approx_pi_reporter.rb
require  File.dirname(__FILE__) + '/approx_pi_finder'

class ApproxPiReporter
  REPEATS = 1000

  attr_reader :sample_size, :best_pi, :worst_pi, :sum_of_pis
  attr_reader :lowest_error, :highest_error

  def initialize(sample_size)
    @sample_size = sample_size
  end

  def generate
    approx_pi, error           = analyze

    @best_pi,   @lowest_error  = approx_pi, error if error.abs < lowest_error.abs
    @worst_pi,  @highest_error = approx_pi, error if error.abs > highest_error.abs
    @sum_of_pis                = sum_of_pis + approx_pi
  end

  def repeat_generates
    @lowest_error  = 100.0
    @highest_error = 0.0
    @sum_of_pis    = 0.0
    1.upto(REPEATS) { generate }
  end

  def output
    repeat_generates
    output_to_console
    output_to_files
  end

  def output_to_console
    $stdout.puts "\n"
    $stdout.puts "For sample size of #{sample_size}"
    $stdout.puts "Best PI  : #{best_pi} (error = #{lowest_error}%)"
    $stdout.puts "Worst PI : #{worst_pi} (error = #{highest_error}%)"
    $stdout.puts "Avg. PI  : #{average_pi} (error = #{average_error}%)"
    $stdout.flush
  end

  def output_to_files
    write_to_file "../output/ruby.average.csv", average_error
    write_to_file "../output/ruby.best.csv", lowest_error
    write_to_file "../output/ruby.worst.csv", highest_error
  end

  def average_pi
    sum_of_pis / REPEATS
  end

  def average_error
    (average_pi - Math::PI) * 100.0 / Math::PI
  end

  def analyze
    finder    = ApproxPiFinder.new(sample_size)
    approx_pi = finder.approx_pi
    error     = finder.error

    [approx_pi, error]
  end

  def write_to_file(file_name, property)
    File.open(file_name, "a+") do |f|
      f.puts "#{sample_size},#{'%.4f' % property}\n"
    end
  end
end

File: python/approx_pi_reporter.py
import math
import sys

import approx_pi_finder

class ApproxPiReporter:
  REPEATS = 1000

  def __init__(self, sample_size):
    self.sample_size = sample_size

  def generate(self):
    approx_pi, error = self.analyze()
    if math.fabs(error) < math.fabs(self.lowest_error):  self.best_pi,   self.lowest_error  = approx_pi, error
    if math.fabs(error) > math.fabs(self.highest_error): self.worst_pi,  self.highest_error = approx_pi, error
    self.sum_of_pis = self.sum_of_pis + approx_pi

  def repeat_generates(self):
    self.lowest_error  = 100.0
    self.highest_error = 0.0
    self.sum_of_pis    = 0.0
    for _ in range(self.REPEATS):
      self.generate()

  def output(self):
    self.repeat_generates()
    self.output_to_console()
    self.output_to_files()

  def output_to_console(self):
    print("\n")
    print("For sample size of {0}".format(self.sample_size))
    print("Best PI  : {0} (error = {1}%)".format(self.best_pi, self.lowest_error))
    print("Worst PI : {0} (error = {1}%)".format(self.worst_pi, self.highest_error))
    print("Avg. PI  : {0} (error = {1}%)".format(self.average_pi(), self.average_error()))
    sys.stdout.flush()

  def output_to_files(self):
    self.write_to_file("../output/python.average.csv", self.average_error())
    self.write_to_file("../output/python.best.csv", self.lowest_error)
    self.write_to_file("../output/python.worst.csv", self.highest_error)

  def average_pi(self):
    return self.sum_of_pis / self.REPEATS

  def average_error(self):
    return (self.average_pi() - math.pi) * 100.0 / math.pi

  def analyze(self):
    finder    = approx_pi_finder.ApproxPiFinder(self.sample_size)
    approx_pi = finder.approx_pi()
    error     = finder.error()
    return [approx_pi, error]

  def write_to_file(self, file_name, property):
    with open(file_name, mode='a+', encoding='utf-8') as file:
      file.write("{0},{1:.4f}\n".format(self.sample_size, property))

